### PR TITLE
Use postcss-selector-parser for selector handling

### PR DESCRIPTION
**Pull request by jacobp100**

### CODE REVIEW

**Comment by jacobp100** [comment]
This broke one test: *should not find components with invalid attributes* (i.e. `[-foo]` won't match anything).

I'm not sure what's best here. If we want to keep this behaviour, it'll have to be an extra check on attribute selectors. I feel like selecting invalid JSX attributes should throw.

**Comment by Aweary** [comment]
Very nice, I was [just mentioning](https://github.com/airbnb/enzyme/issues/439#issuecomment-225700390) doing something like this here. Do you know if this would also resolve https://github.com/airbnb/enzyme/issues/456?

ping @ljharb what do you think?

File: src/ComplexSelector.js [hunk 1]
```diff
@@ -7,31 +7,31 @@ export default class ComplexSelector {
...
 
-      list.push(sel);
       return list;
     }, []);
   }
 
   handleSelectors(selectors, wrapper) {
     const recurseSelector = (offset, fn, pre) => {
-      const predicate = pre || this.buildPredicate(selectors[offset]);
+      const predicate = pre || this.buildPredicate(selectors[offset].toString());
```
Comment (ljharb) [inline src/ComplexSelector.js]: `toString` is not guaranteed to return a String - please wrap in `String()` instead, throughout.

File: src/Utils.js [hunk 2]
```diff
@@ -161,13 +162,14 @@ export function withSetStateAllowed(fn) {
   }
 }
 
-export function splitSelector(selector) {
-  return selector.split(/(?=\.|\[.*\])|(?=#|\[#.*\])/);
+export function getAst(selector) {
```
Comment (ljharb) [inline src/Utils.js]: This name is pretty broad for just dealing with CSS selectors - maybe `getSelectorAST`?

File: src/Utils.js [hunk 3]
```diff
@@ -161,13 +162,14 @@ export function withSetStateAllowed(fn) {
...
+  let value;
+  parser(ast => { value = ast; }).process(selector);
+  return value;
 }
 
-export function isSimpleSelector(selector) {
-  // any of these characters pretty much guarantee it's a complex selector
-  return !/[~\s:>]/.test(selector);
+export function isValidPropName(propName) {
+  return /^[$A-Z_][0-9A-Z_$-:]*$/i.test(propName);
```
Comment (ljharb) [inline src/Utils.js]: Rather than using `/i`, could we just specify `a-zA-Z`? I think it's equally readable but would be more efficient.

**Comment by ljharb** [comment]
Made a few comments - overall I think this looks awesome! @jacobp100, could you add more shallow and mount tests that cover a number of expected-to-work selectors, so that things don't break without us knowing in the future?

**Comment by Aweary** [comment]
@jacobp100 I mentioned a couple of valid selectors that are broken in [this comment](https://github.com/airbnb/enzyme/issues/413#issuecomment-221636721). If we could test those as well that'd be great.

File: src/ShallowTraversal.js [hunk 4]
```diff
@@ -108,34 +107,34 @@ export function nodeMatchesObjectProps(node, props) {
   return isSubset(propsOfNode(node), props);
 }
 
+function buildSelectorPredicate(selector) {
+  const nodes = getAst(selector).nodes[0];
+  return AND(nodes.map(node => {
+    switch (node.type) {
+      case 'class':
```
Comment (blainekasten) [inline src/ShallowTraversal.js]: It's a bit nit, but the old way of using object constants seemed safer/nicer than just strings. Can you update this to use constants?

**Comment by blainekasten** [comment]
This is awesome and exciting. Thanks for doing this work @jacobp100

File: src/ShallowTraversal.js (hunk 4 above)
Comment (jacobp100) [inline src/ShallowTraversal.js]: The strings come from `postcss-selector-parser`, and aren't exported. But I could definitely put a map in util if you think that's best!

Comment (blainekasten) [inline src/ShallowTraversal.js]: Yeah. Just put it in a map so there is a central place of seeing the node types supported

Comment (ljharb) [inline src/ShallowTraversal.js]: It would also be great to file an issue or send a PR upstream to get `postcss-selector-parser` to export the strings.

**Comment by ljharb** [comment]
This LGTM! @lelandrichardson, any final thoughts before we merge?

**Comment by Aweary** [comment]
For reference, I've opened a PR upstream to export the selector types: https://github.com/postcss/postcss-selector-parser/pull/70

**Comment by lelandrichardson** [comment]
Really liking this direction.... Thanks for this work @jacobp100 . I have a couple of questions though:

1. I gather we are ignoring combined selectors (ie, `.foo, .bar`) in some situations (or is it all situations?). In this case, we should probably throw a warning or something to the developer explaining that only the first one is being used, or else it might be misleading.

2. are we no longer warning at all for unsupported CSS selectors? It seems there are still a number left.

3. this gets an AST for the CSS selectors (which is great!) but leaves traversal up to us. For descendant selectors etc. this can be quite hairy. I don't seem to see a lot of tests covering this behavior? It seems important to me if we are adding descendant (and other nonlocal selectors like sibling, direct descendant, etc.) then we will really want to make sure we get this right. 

Sorry if I'm missing something... but I feel like this needs to be merged in with a little bit more care. I love the direction and it seems like this will be much more robust than attempting the parsing ourselves, but let's make sure we understand the change we are making, and ensure that any functionality that we are adding ends up working as close to the CSS spec as it can be, or if it differs, documenting the differences.

Does that seem reasonable?

**Comment by jacobp100** [comment]
Agreed. I think it would make sense to throw for multiple selectors.

I think I throw for all invalid selectors. Can you see a case where I haven't?

I'm actually looking into traversals in another branch. It's a bigger undertaking though.

**Comment by Aweary** [comment]
One concern I have is bringing in a dependency that isn't being actively maintained. Looking at `postcss-selector-parser` it doesn't seem like they are actively maintaining it at the moment. The last commit was almost a month ago and there are PRs that haven't been addressed for about the same amount of time.

If we make this a dependency and run into upstream issues we may be facing an uphill battle getting those fixed. They could just be busy right now, but it's worth considering.

**Comment by ljharb** [comment]
If it's abandoned that's one thing, but we can always fork it and publish our own in that case - I'm not too concerned about anything that's in a git repo.

**Comment by lelandrichardson** [comment]
An alternative that's been brought up before is this: https://github.com/jquense/bill which is used in https://github.com/jquense/teaspoon

cc @jquense to see if he has anything to add on this thread!

**Comment by jquense** [comment]
I've suggested bill before as a possibility. i'd be happy to have it consumed by more testing libraries, I'd also be happy to have more collaborators which may be a better option to forking something :).  bill solid at this point and supports back to react13

**Comment by Aweary** [comment]
@jquense coming back to this, it looks like `bill` would take care of traversal for us right? I would be open to trying to implement this with `bill` instead so we don't have to include a bunch of AST traversal logic here too.

**Comment by jquense** [comment]
@Aweary ya `bill`, exposes a bunch of traversal utilities, and abstracts out the differences between element and instance trees so that you can use the same traversal logic (where that makes sense) for both. High level it wraps matched trees in `Node` objects that expose DOM Element-esque traversal properties `prevSibling`, `children`, `parentNode`, etc; [it makes implementing a wide range of selectors nice and easy](https://github.com/jquense/bill/blob/master/src/selectors.js)

feel free to ping me where ever if you have any questions/concerns 👍

**Comment by aweary** [comment]
@jacobp100 is there any chance you still want to work on this?

**Comment by jacobp100** [comment]
I think this pr should be complete. But let me know if you have any bugs!

**Comment by ljharb** [comment]
@jacobp100 at the least it'd need a fresh rebase - do you mind doing that?

**Comment by aweary** [comment]
@jacobp100 would you mind rebasing and resolving the merge conflicts? If not, no worries. I can try to take this over if you don't have the time. I'd really like to get this merged soon.

**Comment by jquense** [comment]
@aweary you are everywhere man

**Comment by aweary** [comment]
@jquense so many interesting projects, so little time 😄

**Comment by jquense** [comment]
@aweary I know! any shot of y'all incorporating https://github.com/jquense/bill ? id love to not maintain my other testing library on top of it and just use enzyme :P

**Comment by aweary** [comment]
@jacobp100 At this point Enzyme has been entirely rewritten and this PR has fallen really far out of sync. I'm going to close this in favor of https://github.com/airbnb/enzyme/pull/1086 which uses a custom selector parser that supports the variant selector grammar we expect and utilizes the new RST spec we've defined.

Sorry this never got merged, we really appreciate all the hard work you did on this!

**Pull request closed without merging**